### ISSUE TITLE
Enhance efficiency of remote feature requests

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#6015]
## Feature Description

Currently enabled feature flags are requested from the SK service as part of the filtering of active features but only if none are saved yet. This has the potential to request features more than it should if the request fails, or if the result isn't able to be stored. Once saved, they are only refreshed twice a day using a WP cron. Also, if WP cron is disabled, the features would never be updated.

Features shouldn't really be requested more than once per page load, but potentially more limited than that to avoid sending too many requests in some circumstances.

There's also a potential for some minor refactoring here, extracting the functionality around remote features into a new class, breaking out some bits from the our `Authentication` class.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Remote features should not be fetched as part of the `googlesitekit_is_feature_enabled` filter callback regardless of the saved value
* Remote features should be immediately requested and saved during the initial setup, as soon as `Credentials` are set (see `Authentication\Setup::handle_site_code`)
* Remote features should only be requested if connected to the proxy
* Otherwise, feature flag activation via the above filter should only use the saved value when present
* [Comments](https://github.com/google/site-kit-wp/issues/6129) and [workarounds](https://github.com/google/site-kit-wp/blob/d92a77704439ec02ee828929837bc0447c39521d/includes/Core/Authentication/Authentication.php?rgh-link-date=2022-11-09T19%3A13%3A50Z#L1410-L1414) related to fetching features before modules are registered should be irrelevant now and should be removed
* All functionality around remote features should be extracted into one or more new classes, the first (perhaps only) being `Core\Util\Remote_Features`
* Relevant test coverage should be added to cover the behavior and any new classes

## Implementation Brief

* Add `Core\Util\Remote_Features` class, 
  * Declare the class as `final`
  * Utilize `Method_Proxy_Trait` trait
  * Declare `private` variables `authentication`, `credentials`, `google_proxy`, and `options` to hold instances of `Authentication`, `Credentials`, `Google_Proxy` and `Options`, respectively
  * Add `constructor` method, 
    * Accept `Authentication`,  and `Options` (optional) instances as parameters
    * Assign the `Authentication` instance and `Options` (if provided) to the previously declared respective variables. If `Options` is not passed, initialize a new instance.
    * Obtain and assign instances of `Credentials` and `Google_Proxy` from the `Authentication` instance as illustrated [here](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Core/Authentication/Setup.php#L101-L102)
  * Migrate these methods from the `Authentication` class to the new `Remote_Features` class
    * `filter_features_via_proxy` (but exclude the [feature fetching](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Core/Authentication/Authentication.php#L1402)
    * `fetch_remote_features` (should be made `public`)
    * `cron_update_remote_features`  
    **(additionally return early if `! credentials.using_proxy`)**
  * Add `register` method
    * Migrate the logic found [here](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Core/Authentication/Authentication.php#L283-L288) from the `Authentication` class 
* Update `Authentication` class
  * Remove the methods and hooks that were transferred to the `Remote_Features` class
* Update `includes/Core/Authentication/Setup.php`
  * Declare new `private` variable named `remote_features`
  * In the `__construct` method add `Remote_Features` instance to the list of current parameters and assign it to the `remote_features` variable
  * In `handle_site_code` method,  invoke the `fetch_remote_features` from the `remote_features` instance right after the credentials [are assigned](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Core/Authentication/Setup.php#L371)
* Modify `includes/Plugin.php`
  * After `Authenticate` is [instantiated](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Plugin.php#L166-L167), instantiate `Remote_Features` in a variable named `remote_features` and invoke its `register` method
  * Supply the `remote_features` variable to the `Setup` [instantiation process](https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/includes/Plugin.php#L192) 
* Comments and workarounds mentioned in AC, seems to be already removed (not present at the time of writing).

### Test Coverage

* Add tests for `Remote_Features` class
* Update `tests/phpunit/integration/Core/Authentication/Setup_Test.php` and `tests/phpunit/integration/Core/Authentication/AuthenticationTest.php` tests 

## QA Brief

* Disconnect Site Kit and delete the `googlesitekitpersistent_remote_features` record in the `wp_options` table.
* Now set up Site Kit and ensure the above record is created in the database as soon as the plugin is connected.
* Delete the record again and refresh the plugin (visiting other pages as well). Ensure the record is not created anymore in the database.

## Changelog entry

* Prevent excessive requests for remote feature flags.

**Comment by jimmymadon** [comment]
@aaemnnosttv Just wanted to clarify:

>if WP cron is disabled, the features would never be updated.

Is there anything in the AC which acts on this? Should we be fetching remote features once, and only once, on _every_ page load, even if the request was successful? I know the AC says "once per page load at most". In which case shouldn't we? if they exist in the persistent DB option, should we still skip the fetch? If yes, then we will still be relying on the cron to refresh them.

**Comment by jimmymadon** [comment]
@FlicHollis As discussed in the standup, un-assigning myself from this one as I'll focus on the Key Metrics widget issues.

**Comment by FlicHollis** [comment]
Perfect thanks so much @jimmymadon !

**Comment by bethanylang** [comment]
@jimmymadon I know you've been swamped with KMW stuff, so please unassign this one if you want someone else to pick up.   :)

**Comment by aaemnnosttv** [comment]
Thanks @bethanylang – It's not highly urgent but would be great to get moving in this sprint.

**Comment by zutigrm** [comment]
@aaemnnosttv I have a question regarding this point from AC:

> All functionality around remote features should be extracted into one or more new classes, the first (perhaps only) being Core\Util\Remote_Features

Currently all feature flags, include wrapping several different bits and pieces of functionality in different files. Is this still relevant in regarding to moving all these under separate class? Just thinking, when feature flag is removed we would move back the code blocks from this class to where it is added via different methods, vs simply deleting the feature flag wrapper in files. What do you think?

**Comment by aaemnnosttv** [comment]
@zutigrm this issue isn't intended to change anything about how feature flags are _used_. The point in the AC here is only about extracting the functionality around fetching/storing/filtering features from the service which is currently tangled up in the `Authentication` class.

**Comment by aaemnnosttv** [comment]
@zutigrm I realize this IB is an older one of yours, but it could use a bit of a refresh to make it a bit more concise and focused on the important details. That probably isn't worth doing though since the gist of the IB is there and covers all the important points. Certain things like making the class `final` I would say aren't necessary.

IB ✅

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

- Tested on dev environment.

@jimmymadon  I followed QAB  below steps -

- > Disconnect Site Kit and delete the googlesitekitpersistent_remote_features record in the wp_options table.
- > Now set up Site Kit and ensure the above record is created in the database as soon as the plugin is connected.

and found that if I simply disconnect site kit and do not reset the plugin and again reconnect then `googlesitekitpersistent_remote_features `record is not getting create in the options table. To recreate this record we require to reconnect site kit and again reconnect. 
Can you please confirm if QAB is correct or this record should get create when we simply disconnect and reconnect the plugin ? 

<details>

**When user disconnect and reconnect the plugin.**


https://github.com/google/site-kit-wp/assets/94359491/759459d6-9245-4633-a11d-a24cc1ba2dd4


**When user disconnect > Reset > Reconnect plugin.**


https://github.com/google/site-kit-wp/assets/94359491/ae2b6b58-ef18-456f-8bee-ce29daa9415f

</details>

**Comment by mohitwp** [comment]
### QA Update ✅ 

Thank you @jimmymadon !

- Tested on dev environment.
- Followed QAB and verified `googlesitekitpersistent_remote_features` record are getting create and not getting create as per scenarios mentioned under QAB.


<details>

**When user disconnect > Reset > Reconnect plugin.**

https://github.com/google/site-kit-wp/assets/94359491/ae2b6b58-ef18-456f-8bee-ce29daa9415f

**Comment by aaemnnosttv** [comment]
⚠️ I noticed a few inconsistencies in the implementation that aren't broken but can be updated and simplified a little so I'm sending this back to CR with a small followup PR.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@aaemnnosttv I have an observation to share.   

I set up a new site and analytics account.  After connecting, I followed the QAB: 1) Disconnected (_not reset_) Site Kit and deleted the `googlesitekitpersistent_remote_features` record in the `wp_options` table. I then reconnected Site Kit, and all was fine.  The site is in gathering data state.  I then went into the database and noticed that the `googlesitekitpersistent_remote_features` record in the `wp_options` table does not appear. 

I refreshed the database to make sure. 

I also ran through the same steps but with a site with live data, and the same issue occurs.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

@jimmymadon thank you. That'll teach me to read the ticket comments! 😄 

- I reset Site Kit and deleted the `googlesitekitpersistent_remote_features` record in the `wp_options` table. Then, set up Site Kit and the above record is created in the database as soon as the plugin is connected.
- I deleted the record again and refreshed the plugin and the record is not created anymore in the database.
